### ISSUE TITLE
[TEST] Op-test framework reports benign bf16↔DLFloat16 reduction rounding (≤1 ULP) as a numerical failure (torch.sum)

### ISSUE THREAD

**Issue by moriohara** [torch-spyre/torch-spyre#4232]
### What happened?

Two model op-test cases fail as "numerical mismatch":

- `test_model_ops_db_torch_sum_1__gpt-oss-20b_yaml__spyre_float16` — bf16 `[11, 4, 2880]` → `sum(dim=1)`
- `test_model_ops_db_torch_sum_2__gpt-oss-20b_yaml__spyre_float16` — bf16 `[ 1, 4, 2880]` → `sum(dim=1)`

(gpt-oss MoE expert reduction, [`moe.py#L171`](https://github.com/huggingface/transformers/blob/v5.15.0/src/transformers/integrations/moe.py#L171): `weighted_out.view(num_tokens, num_top_k, hidden_dim).sum(dim=1)`.)

**These are not a backend compute bug.** The Spyre reduction is numerically correct *for the device's internal float format*; the failure comes entirely from the **test framework comparing a device result (accumulated in DLFloat16, 1s/6e/9m) against a CPU reference accumulated in fp32, at a tolerance (`atol=rtol=0.005`) tighter than one bf16 ULP.** The discrepancy is ≤ 1 bf16 ULP (`0.015625` at values in `[2,4)`) on ~0.8% of elements — exactly what the fp16/DLFloat16-vs-fp32 format difference predicts.

Two experiments (repro below) establish this:

1. **Exact-integer control** — with integer inputs whose 4-wide sums are exactly representable in bf16/fp16/DLFloat16/fp32 alike, Spyre matches the reference **bit-exactly** (0/31680 and 0/2880 mismatched, maxdiff 0). The reduction *logic* is correct; only rounding differs.
2. **Accumulator-width sweep** — simulating the reduction on CPU at different accumulator mantissa widths and comparing bit-for-bit to the Spyre output (`[11,4,2880]`):

   | accumulator | bit-exact vs Spyre |
   |---|---|
   | bf16 (1s/8e/7m) | 23848 / 31680 |
   | **DLFloat16 (1s/6e/9m)** | **29673 / 31680** ← best |
   | fp16 (1s/5e/10m) | 29541 / 31680 |
   | fp32 (1s/8e/23m) | 29013 / 31680 |

   Spyre tracks a **DLFloat16 (9-mantissa) accumulator** — wider than bf16, narrower than fp32 — then rounds to bf16. Against an fp32-accumulated CPU reference that double-rounds ±1 bf16 ULP on a small fraction of elements. This is expected behavior of the device format, not an error.

The framework should distinguish *real* numerical errors from differences that are purely an artifact of the fp16/DLFloat16 internal format.

### What did you expect to happen?

The op-test should **not** fail cases whose only discrepancy is fp16/DLFloat16-vs-fp32 rounding. The framework should exclude that format-rounding noise by adjusting **the error allowance and/or the initial tensor values** — two complementary levers:

1. **Adjust the error allowance.** Compare bf16/fp16 reduction (and cast) outputs with a tolerance that admits the device format's rounding — a per-dtype/per-op floor of ≈1 ULP of the output dtype, or a tolerance scaled by reduction width — instead of the fixed `atol=rtol=0.005`. Equivalently, compare against a **format-aware reference** (accumulate/round the reference at DLFloat16 precision) rather than an fp32 CPU accumulation. The same drift class is already **xfailed** in the inductor op-tests under the looser `atol=rtol=0.1` (see below); the model op-test DB just applies a much tighter fixed tolerance and turns it into a hard failure.

2. **Adjust the initial tensor values.** Generate inputs whose intermediate/partial results are exactly representable in the device format (bf16/DLFloat16), so format rounding cannot contribute to the comparison at all. The exact-integer control in the repro demonstrates this: with such inputs the device is **bit-exact** (0 mismatches), leaving any remaining diff attributable only to a real bug. This keeps a tight tolerance meaningful for reductions/casts.

**Important:** the goal is to stop flagging *format-rounding* artifacts — **not** to stop detecting numerical errors in general. Whichever lever is used, the framework must still catch genuine correctness bugs (wrong tensor-layout interpretation, incorrect reduction axis/order, stickification/relayout corruption, dropped or double-counted elements, etc.). Those manifest as deviations far larger than one ULP and/or across many elements, so a format-aware tolerance (≈1 ULP for the output dtype) or exactly-representable inputs still surface them while ignoring benign fp16/DLFloat16 rounding. The bar to raise, not lower: distinguish "off by the last representable bit of the device format" from "off because the computation is wrong."

### How can we reproduce it?

```python
import torch
import torch_spyre

@torch.compile(dynamic=False)
def sum_dim1(x):
    return x.sum(dim=1)

# (a) random bf16 -> "fails" only by <=1 ULP
torch.manual_seed(123)
x = torch.rand(11, 4, 2880, dtype=torch.bfloat16)
ref = x.sum(dim=1)                              # CPU eager (fp32 accumulation)
out = sum_dim1(x.to("spyre")).cpu()
print("max abs diff:", (out.float()-ref.float()).abs().max().item())   # 0.015625 (1 bf16 ULP)

# (b) exact-integer control -> device is bit-exact (no format rounding needed)
xi = torch.randint(0, 16, (11, 4, 2880)).to(torch.bfloat16)
oi = sum_dim1(xi.to("spyre")).cpu()
print("int-exact bit-exact:", torch.equal(oi.float(), xi.sum(dim=1).float()))  # True
```

### Any environmental details we need to know?

- torch-spyre: observed on `e97074b` (2026-09-01 run) and still on `8ef9312` (2026-09-02)
- torch `~=2.13.0`, transformers `v5.15.0`
- Single Spyre VFIO device; dev env sourced via `torch-spyre-docs/scripts/dev-env.sh`

### Anything else we need to know?

Prior art in the test suite — the **same marginal reduction-drift class**, handled as xfail under the inductor tolerance (`atol=rtol=0.1`, `tests/inductor/test_inductor_ops.py`):
- `test_slice_stick_reduce_dim1[sum_3d128_01]` — *"the fp16 sum-reduction … drifts a single element by ~0.11 (>0.1 tol) … Same class as the cases xfailed in commit 3a2d482. amax on the same shape passes, so target sum only."*
- `test_sum_keepdim1[fp32_3d_dim_neg1]`
- `tests/inductor/sum.py::test_roundtrip_sum` sidesteps it entirely by accumulating in fp32 (`x.to(fp32).sum().to(fp16)`) with a loose `0.5` tolerance.

Scope: this is likely **not unique to `torch.sum`** — other `#3083` cases currently labeled "Numerical mismatch" on reduction/cast ops (e.g. the fp16↔fp32 conversion family) may be the same false-positive class. A consistent framework tolerance policy (format-aware reference or per-dtype/per-op tolerance) would resolve the class rather than one op at a time. If instead the intent is bit-accurate parity, that requires the backend to accumulate reductions in fp32 before the bf16 write (cf. #1871, #2534) — but that is a backend feature request, distinct from this framework false-positive.

### Relevant log output

```Shell
[sum_1__gpt-oss-20b]  rand       bf16 [11,4,2880]  match(tol)=False  maxdiff=0.015625  242/31680
[sum_1__gpt-oss-20b]  int-exact  bf16 [11,4,2880]  match(tol)=True   maxdiff=0.000000    0/31680
[sum_2__gpt-oss-20b]  rand       bf16 [ 1,4,2880]  match(tol)=False  maxdiff=0.015625   24/2880
[sum_2__gpt-oss-20b]  int-exact  bf16 [ 1,4,2880]  match(tol)=True   maxdiff=0.000000    0/2880

accumulator sweep (bit-exact vs Spyre, [11,4,2880]):
  bf16   (1/8/7)   23848/31680
  DLF16  (1/6/9)   29673/31680   <- best
  fp16   (1/5/10)  29541/31680
  fp32   (1/8/23)  29013/31680
```
